elastic/kibana
# [Log Explorer] Add Explorer app locator

closes https://github.com/elastic/kibana/issues/164995
closes https://github.com/elastic/kibana/issues/165618
closes https://github.com/elastic/kibana/issues/166596

## 📝  Summary

### Observability Log Explorer Locators:

This PR adds 2 new customized locators to the Observability log explorer profile. At the moment we implemented:

       1- Single dataset selector locator
       2- All dataset selector locator

With more locators to come in the future depending on the use cases.

### Log Explorer Locators:

We also added a log explorer locator that navigates to discover, this can be used in case the **Observability Log Explorer** plugin is disabled.

### Logs Onboarding:

The PR also replaces the temp navigation to the default discover we implemented for[ 8.10 here](https://github.com/elastic/kibana/pull/163218) with the above new Observability Log Explorer locators.

### APM:

After [disabling infra plugin in serverless projects](https://github.com/elastic/kibana/pull/165289), APM links to infra locators in serverless have been replaced to use the above locators.

### Observability Landing Page:

The landing page now redirects to the Log Explorer if `logs-*-*` has data in it, otherwise the flow continues as before.

### Necessary Refactoring:

To avoid the circular dependency between `ObservabilityLogExplorer` & `ObservabilityOnboarding` after each one using the other's locator and importing the necessary types, I moved the type definition for all locators in the `deeplinks` package.

## ✅  Testing

- Onboarding Wizard in Serverless and Stateful

    1. Navigate to the onboarding flow `/app/observabilityOnboarding/`
    2. Choose either System logs or Stream log files
    3. Go through the onboarding wizard
    4. Click the Explore logs button at the end
    5. You should be redirected to observability log explorer with the integration and dataset preselected.

- APM links in Serverless

    1. Navigate to APM and click on the logs links as shown in the Demos below
    2. All links should navigate to Observability Log Explorer with the queries set in the search bar.

## 🎥 Demos

- APM Serverless

https://github.com/elastic/kibana/assets/11225826/7161364e-333f-4ac4-87d5-7f1ffec699b3


- APM Stateful

https://github.com/elastic/kibana/assets/11225826/058c9587-b766-4d4f-a73d-50fd381be4bb


- Onboarding Serverless


https://github.com/elastic/kibana/assets/11225826/ee1cab42-f91c-4558-aa5f-4fa7e8963427

- Onboarding Stateful


https://github.com/elastic/kibana/assets/11225826/a376a12b-499b-4488-a75a-d06e81f8e21d

- Observability Landing Page 


https://github.com/elastic/kibana/assets/11225826/c1c084ca-b1b1-4c4b-a4e6-ae8e157dcf57
